### PR TITLE
MODLD-374: Convert MARC field 502 to Dissertation resource

### PR DESCRIPTION
**Pull request by pkjacob**

### CODE REVIEW

File: src/test/java/org/folio/marc4ld/mapper/field502/Bibframe2Marc502IT.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+@EnableConfigurationProperties
+@SpringBootTest(classes = SpringTestConfig.class)
+class Bibframe2Marc502IT {
+
+  @Autowired
+  private Bibframe2MarcMapperImpl bibframe2MarcMapper;
+
+  @Test
+  @Disabled // TODO - Discuss with team tomorrow
+  void shouldMapField502() {
```
Comment (pkjacob) [inline src/test/java/org/folio/marc4ld/mapper/field502/Bibframe2Marc502IT.java:39]: Hi @PBobylev @aleksei-pronichev  @askhat-abishev,

I noticed one problem with graph -> marc transformation. 

Let us consider this graph

image

When MARC is derived from this graph, the expected MARC is as below

> 502 $b PHD $c XYZ University
> 502 $b Doctor $c ABC University

However, the actual MARC is as below
> 502 $b PHD
> 502 $c XYZ University
> 502 $b Doctor
> 502 $c ABC University

From this MARC, we can't correlate between a degree and the corresponding granting institution.

Comment (aleksei-pronichev) [reply src/test/java/org/folio/marc4ld/mapper/field502/Bibframe2Marc502IT.java:39]: @pkjacob DataFieldCombiner is ok for me. It was created exactly for this cases;
pluses
1. just need to add 1 class, implemented the interface
2. no changes in global logic

**Review by aleksei-pronichev — APPROVED** [review]

File: src/test/java/org/folio/marc4ld/mapper/field502/Marc2Bibframe502IT.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    var result = marc2BibframeMapper.fromMarcJson(marc);
+
+    // then
+    assertThat(result)
+      .isNotNull()
+      .extracting(this::getDissertationEdge)
+      .satisfies(e -> validateEdge(e, PredicateDictionary.DISSERTATION,
+        List.of(ResourceTypeDictionary.DISSERTATION),
+        Map.of(
+          "http://bibfra.me/vocab/lite/label", List.of("dissertation label"),
```
Comment (askhat-abishev) [inline src/test/java/org/folio/marc4ld/mapper/field502/Marc2Bibframe502IT.java:43]: why not ```LABEL.getValue()```?

**Review by askhat-abishev — APPROVED** [review]

File: src/test/java/org/folio/marc4ld/mapper/field502/Marc2Bibframe502IT.java (hunk 2 above)
Comment (aleksei-pronichev) [reply src/test/java/org/folio/marc4ld/mapper/field502/Marc2Bibframe502IT.java:43]: using part of the code database in tests isn't good practice
It  hides mistakes and doesn't show the expected results

Suggest keeping this practice and don't use classes from code or our main libs

File: src/main/java/org/folio/marc4ld/service/ld2marc/resource/Resource2MarcRecordMapperImpl.java [hunk 3]
```diff
@@ -35,16 +36,15 @@ public class Resource2MarcRecordMapperImpl implements Resource2MarcRecordMapper
...
-  private final DataFieldPostProcessor dataFieldPostProcessor;
   private final Comparator<Subfield> subfieldComparator;
+  private final Supplier<DataFieldPostProcessor> dataFieldPostProcessorSupplier;
 
   @Override
   public Record toMarcRecord(Resource resource) {
     var marcRecord = marcFactory.newRecord();
     var cfb = new ControlFieldsBuilder();
     var dataFields = getFields(new ResourceEdge(null, resource, null), cfb);
-    var combinedFields = dataFieldPostProcessor.apply(dataFields);
```
Comment (pkjacob) [inline src/main/java/org/folio/marc4ld/service/ld2marc/resource/Resource2MarcRecordMapperImpl.java:46]: Here, we were combining the MARC data fields after generating all the MARC fields. At this stage, we don't know which MARC fields should be combined and which should not. So, removed this logic from here

File: src/main/java/org/folio/marc4ld/service/ld2marc/resource/Resource2MarcRecordMapperImpl.java [hunk 4]
```diff
@@ -65,10 +65,14 @@ private List<DataField> getFields(ResourceEdge edge, ControlFieldsBuilder cfb) {
...
     var resource = edge.getTarget();
-    var dataFields = rules.stream()
+    var fieldsFromResource = rules.stream()
       .flatMap(tagToRule -> mapToDataFields(edge, cfb, tagToRule));
-    return Stream.concat(dataFields, getOutgoingEdges(cfb, resource))
-      .toList();
+    var fieldsFromEdges = getOutgoingEdges(cfb, resource);
+    var combinedFields = Stream.concat(fieldsFromResource, fieldsFromEdges).toList();
+
+    return dataFieldPostProcessorSupplier.get()
```
Comment (pkjacob) [inline src/main/java/org/folio/marc4ld/service/ld2marc/resource/Resource2MarcRecordMapperImpl.java:73]: Here, we are working on a graph resource and its outgoing edges. Looking at the type of the current resource & the MARC tag, we can determine whether the MARC records has to be combined. So, moved the combiner logic to here.

We need to get a new `dataFieldPostProcessor` for each resource of the graph. Hence, using a `supplier` here, which will supply a new `dataFieldPostProcessor` for the current resource.

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/DataFieldCombinerFactoryImpl.java [hunk 5]
```diff
@@ -1,23 +1,36 @@
...
-import org.springframework.stereotype.Component;
 
-@Component
 public class DataFieldCombinerFactoryImpl implements DataFieldCombinerFactory {
 
+  private final Map<Predicate<DataFieldCombiner.Context>, DataFieldCombiner> combinerMap;
+
+  public DataFieldCombinerFactoryImpl() {
+    combinerMap = Map.of(
+      c -> c.marcTag().equals("245") && c.resourceTypes().contains(INSTANCE), new TitleCombiner(),
```
Comment (pkjacob) [inline src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/DataFieldCombinerFactoryImpl.java:21]: 245 & 010 has to be combined after generating marc tags for the entire INSTANCE.
502 has to be combined for each DISSERTATION resource.

File: src/test/java/org/folio/marc4ld/mapper/field502/Bibframe2Marc502IT.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    //when
+    var result = bibframe2MarcMapper.toMarcJson(resource);
+
+    // then
+    AssertionsForClassTypes.assertThat(result)
+      .isEqualTo(expectedMarc);
+  }
+
+  private Resource createResourceWith502() {
+    Resource dissertation1 = createDissertation("1");
```
Comment (askhat-abishev) [inline src/test/java/org/folio/marc4ld/mapper/field502/Bibframe2Marc502IT.java:51]: var

**Review by askhat-abishev — APPROVED** [review]

File: src/test/java/org/folio/marc4ld/mapper/field502/Bibframe2Marc502IT.java (hunk 6 above)
Comment (pkjacob) [reply src/test/java/org/folio/marc4ld/mapper/field502/Bibframe2Marc502IT.java]: Done

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/DataFieldCombinerFactoryImpl.java [hunk 7]
```diff
@@ -1,23 +1,36 @@
...
+      c -> c.marcTag().equals("010") && c.resourceTypes().contains(INSTANCE), new LccnCombiner(),
+      c -> c.marcTag().equals("502") && c.resourceTypes().contains(DISSERTATION), new DissertionCombiner()
```
Comment (aleksei-pronichev) [inline src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/DataFieldCombinerFactoryImpl.java:23]: Minor.
For me it's a separate class, like:
```
DissertionCombinerFactory
LccnCombinerFactory
```
with interface
`CombinerFactory implements Predicate<DataFieldCombiner.Context>, Supplier<DataFieldCombiner>` {
    
String getTag() // unnecessary but for optimisation to create Map<String, CombinerFactory>
}

**Review by aleksei-pronichev — APPROVED** [review]

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/DissertionCombiner.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
+package org.folio.marc4ld.service.ld2marc.processing.combine.impl;
+
+
+import java.util.Collection;
+import java.util.Set;
+
+public class DissertionCombiner extends AbstractDataFieldCombiner {
```
Comment (aleksei-pronichev) [inline src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/DissertionCombiner.java:7]: Dissertation

File: src/main/java/org/folio/marc4ld/configuration/DataFieldPostProcessorConfig.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import org.folio.marc4ld.service.ld2marc.processing.combine.impl.DataFieldCombinerFactoryImpl;
+import org.folio.marc4ld.service.ld2marc.processing.impl.DataFieldPostProcessorImpl;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class DataFieldPostProcessorConfig {
+  @Bean
+  public Supplier<DataFieldPostProcessor> dataFieldPostProcessorSupplier() {
+    return () -> new DataFieldPostProcessorImpl(new DataFieldCombinerFactoryImpl());
```
Comment (aleksei-pronichev) [inline src/main/java/org/folio/marc4ld/configuration/DataFieldPostProcessorConfig.java:14]: DataFieldCombinerFactoryImpl is stateless
suggest to remove config and create a class

**Review by aleksei-pronichev — APPROVED** [review]

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/DataFieldCombinerFactoryImpl.java (hunk 7 above)
Comment (pkjacob) [reply src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/DataFieldCombinerFactoryImpl.java]: Thanks @aleksei-pronichev. I merged your commit to this brach.

File: src/main/java/org/folio/marc4ld/configuration/DataFieldPostProcessorConfig.java (hunk 9 above)
Comment (pkjacob) [reply src/main/java/org/folio/marc4ld/configuration/DataFieldPostProcessorConfig.java]: Thanks @aleksei-pronichev. I merged your commit to this brach.

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/DissertionCombiner.java (hunk 8 above)
Comment (pkjacob) [reply src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/DissertionCombiner.java]: Donr

**Pull request merged**